src/peft/tuners/vblora/layer.py

```diff
@@ -0,0 +1,249 @@
+
+        Args:
+            safe_merge (`bool`, *optional*):
+                If True, the merge operation will be performed in a copy of the original weights and check for NaNs
+                before merging the weights. This is useful if you want to check if the merge operation will produce
+                NaNs. Defaults to `False`.
+            adapter_names (`List[str]`, *optional*):
+                The list of adapter names that should be merged. If None, all active adapters will be merged. Defaults
+                to `None`.
+        """
+        adapter_names = check_adapters_to_merge(self, adapter_names)
+        if not adapter_names:
+            # no adapter to merge
+            return
+
+        for active_adapter in adapter_names:
+            if active_adapter in self.vblora_logits_A.keys():
+                base_layer = self.get_base_layer()
+                if safe_merge:
+                    # Note that safe_merge will be slower than the normal merge
+                    # because of the copy operation.
+                    orig_weights = base_layer.weight.data.clone()
+                    orig_weights += self.get_delta_weight(active_adapter)
+                    if not torch.isfinite(orig_weights).all():
+                        raise ValueError(
+                            f"NaNs detected in the merged weights. The adapter {active_adapter} seems to be broken"
+                        )
+                    base_layer.weight.data = orig_weights
+                else:
+                    base_layer.weight.data += self.get_delta_weight(active_adapter)
+                self.merged_adapters.append(active_adapter)
+
+    def unmerge(self) -> None:
+        if not self.merged:
+            warnings.warn("Already unmerged. Nothing to do.")
+            return
+
+        while len(self.merged_adapters) > 0:
+            active_adapter = self.merged_adapters.pop()
+            if active_adapter in self.vblora_logits_A.keys():
+                self.get_base_layer().weight.data -= self.get_delta_weight(active_adapter)
+
+    def _get_low_rank_matrix(self, logits: torch.tensor, vblora_vector_bank, topk) -> torch.Tensor:
+        top_k_logits, indices = logits.topk(topk, dim=-1)
+        topk_weights = F.softmax(top_k_logits, dim=-1)
+        return (topk_weights.unsqueeze(-1) * vblora_vector_bank[indices]).sum(-2)
+
+    def _get_lora_matrices(self, adapter, cast_to_fp32=False) -> Tuple[torch.Tensor, torch.Tensor]:
+        vblora_logits_A = self.vblora_logits_A[adapter]
+        vblora_logits_B = self.vblora_logits_B[adapter]
+
+        # Check for infinity values when training. If found, training was likely resumed from a `save_only_topk_weights` model.
+        if self.training and vblora_logits_A[0, 0].isinf().any():
+            raise RuntimeError(
+                "Found infinity values in logits. Ensure training was not resumed from a `save_only_topk_weights` model."
```

Comment:
Let's add: "**VB-LoRA** logits", otherwise this could be confusing.

Reply:
Added.